File: templates/problems/14_dp_2d/05_target_sum/solution.go
package target_sum

// solveFindTargetSumWays counts ways to assign +/- to reach target. O(n * sum) time, O(sum) space.
func solveFindTargetSumWays(nums []int, target int) int {
	total := 0
	for _, n := range nums {
		total += n
	}
	// We need (total + target) to be even and non-negative
	if (total+target)%2 != 0 || total+target < 0 || total < abs(target) {
		return 0
	}
	subsetSum := (total + target) / 2
	dp := make([]int, subsetSum+1)
	dp[0] = 1
	for _, num := range nums {
		for j := subsetSum; j >= num; j-- {
			dp[j] += dp[j-num]
		}
	}
	return dp[subsetSum]
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

File: templates/problems/14_dp_2d/05_target_sum/typescript/solution.ts
/**
 * Count ways to assign +/- to reach target.
 * Time: O(n * sum), Space: O(sum)
 */
export function solveFindTargetSumWays(nums: number[], target: number): number {
    const total = nums.reduce((acc, n) => acc + n, 0);
    // We need (total + target) to be even and non-negative
    if ((total + target) % 2 !== 0 || total + target < 0 || total < Math.abs(target)) {
        return 0;
    }
    const subsetSum = (total + target) / 2;
    const dp = new Array<number>(subsetSum + 1).fill(0);
    dp[0] = 1;
    for (const num of nums) {
        for (let j = subsetSum; j >= num; j--) {
            dp[j] += dp[j - num];
        }
    }
    return dp[subsetSum];
}
